src/qibolab/instruments/zhinst/sweep.py

```diff
@@ -0,0 +1,139 @@
+"""Pre-execution processing of sweeps."""
+
+from collections.abc import Iterable
+from copy import copy
+
+import laboneq.simple as lo
+
+from qibolab.pulses import Pulse, PulseType
+from qibolab.qubits import Qubit
+from qibolab.sweeper import Parameter, Sweeper
+
+from .util import NANO_TO_SECONDS, measure_channel_name
+
+
+def classify_sweepers(
+    sweepers: Iterable[Sweeper],
+) -> tuple[list[Sweeper], list[Sweeper]]:
+    """Divide sweepers into two lists: 1. sweeps that can be done in the laboneq near-time sweep loop, 2. sweeps that
+    can be done in real-time (i.e. on hardware)"""
+    nt_sweepers, rt_sweepers = [], []
+    for sweeper in sweepers:
+        if (
+            sweeper.parameter is Parameter.amplitude
+            and sweeper.pulses[0].type is PulseType.READOUT
+        ):
+            nt_sweepers.append(sweeper)
+        elif sweeper.parameter is Parameter.bias:
+            nt_sweepers.append(sweeper)
+        else:
+            rt_sweepers.append(sweeper)
+    return nt_sweepers, rt_sweepers
+
+
+class ProcessedSweeps:
+    """Data type that centralizes and allows extracting information about given
+    sweeps."""
+
+    def __init__(self, sweepers: Iterable[Sweeper], qubits: dict[str, Qubit]):
+        pulse_sweeps = []
+        channel_sweeps = []
+        parallel_sweeps = []
+        for sweeper in sweepers:
+            for pulse in sweeper.pulses or []:
```

Comment:
> do not rely on falsiness of None, it could be misleading

why is it misleading? It is a feature of the language and also helps make code compact

Reply:
*Problem with falsiness:* If by any chance `sweeper.pulses` happens to be `0` or `False`, this would result in skipping the iteration instead of a suitable error.

*Problem with loop:* it relies on the priority of `for ... in` and `or` (ok, you could add parenthesis, but it makes many things happening in a single place).

However, the best solution could be to initialize `sweeper.pulses` to `[]` instead of `None`. No pulses is equivalent to an empty list, there should be no need to mark explicitly the case in which it is unset by the user (and in a dataclass you can use a `default_factory()` for that).
Reducing the variability of the type always helps to smoothly account for limit cases.